junit-vintage-engine/src/main/java/org/junit/vintage/engine/discovery/VintageDiscoverer.java

```diff
@@ -37,44 +39,37 @@
 public class VintageDiscoverer {
 
 	private static final IsPotentialJUnit4TestClass isPotentialJUnit4TestClass = new IsPotentialJUnit4TestClass();
-	private final Logger logger;
+	private final CompleteTestClassesResolver completeTestClassesResolver;
+	private final FilteredTestClassesResolver filteredTestClassesResolver;
 	private final TestClassRequestResolver resolver;
+	private final Logger logger;
 
 	public VintageDiscoverer(Logger logger) {
+		this.completeTestClassesResolver = new CompleteTestClassesResolver();
+		this.filteredTestClassesResolver = new FilteredTestClassesResolver(logger);
 		this.logger = logger;
 		this.resolver = new TestClassRequestResolver(logger);
 	}
 
 	public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
 		EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "JUnit Vintage");
 		// @formatter:off
-		collectTestClasses(discoveryRequest)
-				.toRequests()
+		collectTestClassRequests(discoveryRequest)
 				.map(request -> resolver.createRunnerTestDescriptor(request, uniqueId))
 				.filter(Objects::nonNull)
 				.forEach(engineDescriptor::addChild);
 		// @formatter:on
 		return engineDescriptor;
 	}
 
-	private TestClassCollector collectTestClasses(EngineDiscoveryRequest discoveryRequest) {
+	private Stream<TestClassRequest> collectTestClassRequests(EngineDiscoveryRequest discoveryRequest) {
 		Predicate<Class<?>> classFilter = createTestClassPredicate(discoveryRequest);
-		TestClassCollector collector = new TestClassCollector();
-		for (DiscoverySelectorResolver selectorResolver : getAllDiscoverySelectorResolvers(discoveryRequest)) {
-			selectorResolver.resolve(discoveryRequest, classFilter, collector);
-		}
-		return collector;
-	}
-
-	private List<DiscoverySelectorResolver> getAllDiscoverySelectorResolvers(EngineDiscoveryRequest request) {
-		Predicate<String> classNamePredicate = buildClassNamePredicate(request);
-		return asList( //
-			new ClasspathRootSelectorResolver(classNamePredicate), //
-			new PackageNameSelectorResolver(classNamePredicate), //
-			new ClassSelectorResolver(), //
-			new MethodSelectorResolver(), //
-			new UniqueIdSelectorResolver(logger)//
-		);
+		Set<Class<?>> completeTestClasses = completeTestClassesResolver.resolve(discoveryRequest, classFilter);
+		Map<Class<?>, List<RunnerTestDescriptorAwareFilter>> filteredTestClasses = filteredTestClassesResolver.resolve(
+			discoveryRequest,
+			//no filters should be applied for test classes that are requested to be tested completely
```

Comment:
I don't understand this comment. The `classFilter` is applied, isn't it?

Reply:
I also think the next line states absolutely clear what will get filtered and what not. No need for a comment.

Reply:
I was not sure whether this line of code can be understood. @marcphilipp Do you think we need a (a better) comment or should I delete it.

Reply:
Maybe "only include filtered classes that are not already included completely" or something along those lines?

Reply:
I replaced my comment with yours. It is much easier to understand. Thanks!